### ISSUE TITLE
Constraints blowing up

### ISSUE THREAD

**Issue by jlmaccal** [openmm/openmm#3603]
We're working on finalizing our implementation of the Martini CG force field in OpenMM, but are running into problems with the topology for cholesterol.

In Martini, four atoms in cholesterol are constrained in a diamond pattern with five constraints. I have constructed a minimum failing example below. LINCS can handle this construction fine, but OpenMM's constraint algorithm blows up.

```python
import openmm as mm
import openmm.unit as u
import numpy as np


system = mm.System()
system.addParticle(50)
system.addParticle(50)
system.addParticle(50)
system.addParticle(50)

# Add constraints
system.addConstraint(0, 1, 0.4904)
system.addConstraint(0, 2, 0.6019)
system.addConstraint(1, 2, 0.2719)
system.addConstraint(3, 1, 0.7237)
system.addConstraint(3, 2, 0.5376)

platform = mm.Platform.getPlatformByName("Reference")
integrator = mm.LangevinIntegrator(
    300 * u.kelvin, 1 / u.picosecond, 0.002 * u.picoseconds
)
context = mm.Context(system, integrator, platform)

positions = np.array(
    [
        [5.566, 7.728, 3.213],
        [5.494, 7.973, 3.631],
        [5.354, 7.788, 3.773],
        [5.295, 7.811, 4.307],
    ],
    dtype=float,
)
context.setPositions(positions)
context.applyConstraints(1e-4)
new_positions = context.getState(getPositions=True).getPositions(asNumpy=True)
print(new_positions)
```

The OpenMM algorithm appears to behave well up to a tolerance of `1e-2`, but blows up for tighter tolerances.

**Comment by peastman** [comment]
This arrangement of constraints seems to be ill-conditioned, though it isn't immediately obvious to me why.  If you remove any one constraint, it works correctly.  I can sort of see why it might be: each of the triangles 0-1-2 and 1-2-3 on its own would uniquely determine the positions of atoms 1 and 2.  Enforcing both triangles at once can therefore cause them to interfere with each other.  That's not to say the constraints are fundamentally in conflict: there certainly should exist a solution.  But reaching it may be numerically difficult.

If I recall correctly, LINCS cheats and only ever does two iterations?  It doesn't try to iterate constraints to convergence.  Watching the constraint solver in OpenMM, it isn't until the third iteration that it starts to show serious signs of blowing up.

**Comment by jlmaccal** [comment]
To be honest, the details of constraint algorithms are something I've never really dug into.

Yes, LINCS only uses two steps. I'm actually surprised it works on a system like this, as my understanding was that LINCS did not perform well on highly coupled constraint systems.

Any thoughts on how to proceed with CCMA? Would it be possible to apply a preconditioner to make numerical solution more reliable?

**Comment by peastman** [comment]
Maybe?  I think the first step would be to analyze the constraint matrix.  Try calculating its condition number to verify whether ill conditioning is really the problem.  It gets built in ReferenceCCMAAlgorithm.cpp.  You can print out the elements and analyze it in another program.

**Comment by peastman** [comment]
I can get it to converge if I add enough damping to the iteration.  Here's where it updates the positions after each iteration:

https://github.com/openmm/openmm/blob/109f6b2535da4e0c0dd88007d6ca06b4add2ce81/platforms/reference/src/SimTKReference/ReferenceCCMAAlgorithm.cpp#L311-L313

If I multiply `dr` by 0.5 it converges, though it takes 22 iterations.  It seems we're just a bit outside some stability limit.

**Comment by jlmaccal** [comment]
Is it reasonble to make something like that as a default in the name of improved stability in difficult cases, presumably at the expense of slow convergence in more typical cases? Or to allow access to this scaling factor as a parameter? 

Or, is there another approach that could be implemented without a loss of efficiency? I know very little about techniques for preconditioning, but it strikes me that one thing in our favour in this case is the fact that the matrix is static with CCMA, so even if there is a cost to computing the preconditioning it would only have to paid once.

**Comment by peastman** [comment]
The damping is just a hack to help diagnose the problem.  It wouldn't be a good solution since it makes the typical case much slower.  And there's no guarantee it would work in all similar cases.  Really we need to understand better what's going on.  Why does the Newton iteration fail to converge?

**Comment by jlmaccal** [comment]
I translated the C++ code into python to make it a bit easier to experiment. The solution is definitely overshooting/oscillating, which causes it to blow up eventually for this particular geometry.

I've also been looking at the code, and there are a few things that don't make sense to me.

The code first calculates the relative deviation of squares in each constraint `(d_target**2 - d_actual**2) / d_target**2`. This is then multiplied by the inverse Jacobian. This doesn't seem right to me. Wouldn't the Jacobian for that deviation of squares be different than the one for sigma from the paper?

As an example, assume we have two co-linear constraints(so `cos(theta) == 1`) sharing the middle atom. Assume that all masses are the same. The Jacobian for sigma from the paper says that a change $\delta$ to one constraint will lead to a change $0.5 \delta$ to the other. However, for the relative deviation above, this is wrong unless the two constraints have the same lengths. A 5% change in one constraint could result in a larger or smaller change in the other, depending on their relative lengths.

Then later on, when the coordinates are updated, the relative inverse jacobian times relative deviation of squares is used to scale the interatomic vector to derive the correction. However, this doesn't seem quite right to me either. If the error in the relative deviation of squares is 1 percent, shouldn't the correction to the bond length be `sqrt(0.01) * r_ij`?

I'm probably not fully understanding the implementation. But my confusion basically comes from not understanding how the math from the paper can be implemented without ever taking a square root.

**Comment by peastman** [comment]
I'll take a look and see if I can figure it out.  It's been a very long time since I wrote that code.

**Comment by peastman** [comment]
I can't figure out what exactly that code is doing.  But for the most part, it seems to work!  Very confusing.

I rewrote it in the way that made more sense to me.  Your example now converges.  Other examples I tested also still converge, but many of them now take more iterations than they did before.  So did I make it better?  Worse?  I'm not sure!  I'll clean it up and make a PR.  Then perhaps you can look it over and see what you think.

**Comment by peastman** [comment]
> If the code doesn’t find an angle or triangle, then it defaults to zero, but think it might be better to raise an error in this case.

We can't do that.  There are lots of cases where there wouldn't be an angle or triangle.  A trivial example would be a freely jointed chain.

> That have the Jacobian element calculated as zero, but this is clearly incorrect as these constraints share an atom and are coupled.

That's just part of the approximation.  The corresponding element of the true Jacobian depends on the angle formed by the two constraints.  It can be anything from -1 (if they form a straight line) to 1 (if they're on top of each other).  In your case a wide range values are possible, depending on how the molecule folds along its middle.  Since we have no information about what the actual element is likely to be, we set it to 0.

**Comment by jlmaccal** [comment]
Sure, I get the approximation. I’m just wondering if the absence of these couplings from the Jacobian, combined with the otherwise cyclic / highly coupled nature of this particular system, might be leading to the blow up that is being observed.

**Comment by jchodera** [comment]
@peastman: Should this be labeled as a `bug` that should be fixed before the next release?

**Comment by jlmaccal** [comment]
It seems that gromacs also has some [problems](https://pubs.acs.org/doi/pdf/10.1021/acs.jpcb.1c03665) with this specific topology. I understand that the cholesterol parameters for Martini 3 are under development and having more stable constraints is one of the considerations.

**Comment by jlmaccal** [comment]
I played around with adding two dummy angles to specify the missing components of the Jacobian. Cholesterol is approximately planar, so these angles are well defined. I just set the force constant to zero.

With these additional angles, the constraints appear to converge as expected.

One solution to our specific problem might be for us to special-case cholesterol in our parser to add in these extra angles. However, I'm a bit worried about the problem becoming near singular with so many coupled constraints.

**Comment by peastman** [comment]
Near-planar configurations are always ill-conditioned for constraints.  That's because it can only apply constraint forces along constrained directions, and those all lie approximately in the plane.  To leading order, it has no way of applying a force to keep an atom from moving out of the plane.

Adding dummy angles shouldn't cause any problems I can think of.  You aren't increasing the number of constraints, so there's no danger of making it overdetermined.  You're just helping it build a better approximation to the constraint matrix.

**Comment by jlmaccal** [comment]
Regarding planarity, I just meant that this topology has impropers that keep the "diamond" of constrained atoms relatively planar. This allows me to infer what the missing angles is and gives assurance that the actual angle will be close to this. Without the impropers the angle really could be anywhere form 0 to 180 degrees.

I'm still confused about some of the details of the code though. It obviously works and converges faster than a more naive implementation, I just don't understand why.

**Comment by jchodera** [comment]
Did we ever decide if there was a bug in the current code that needed to be addressed, or some further improvements to stability that were required to make constraints more robust?

**Comment by peastman** [comment]
I don't think there's any bug.  It's less clear whether there's a generic way to improve robustness.  In this case, adding some angle terms with zero force constants helped it to build a better approximation to the constraint matrix, which improved convergence.  Could we have gotten that same benefit in a more automated way?  I don't know.

**Comment by jchodera** [comment]
> I don't think there's any bug.

Didn't @jlmaccal and you both verify the implementation was giving numerically and behaviorally distinct results from the expressions in the CCMA paper?

**Comment by peastman** [comment]
No, we just verified that we don't understand the code. :)  But the results are correct, and when I tried to rewrite it in what seemed a more obvious way, it no longer worked!

**Comment by jlmaccal** [comment]
To add some follow up.

In our martini parser, I added code that creates an additional "dummy" dihedral with zero force constant for the special case of cholesterol. The resulting inverse Jacobian no longer blows up and converges on systems containing hundreds of cholesterols.

However, the solution is not always the same as with LINCS. In rare cases, the atom positions can differ by as much as 0.1 nm, which seems larger than I would have expected. These deviations occur when the cholesterol deviates substantially from planar (which is rare due to impropers) which leads to the dummy angle being less correct.

I don't know enough of the math of behind constraint algorithms. Is the solution always unique?

**Comment by peastman** [comment]
Aside from chirality, I believe the solution should always be unique.  Certainly for your cholesterol model, which only has one internal degree of freedom.  For each of the two solvers, try computing the actual distances between the final positions it gives you.  How well do they match the target distances specified by the constraints?

**Comment by peastman** [comment]
> I suspect the issue is that this cholesterol conformation is particularly distorted, such that the specified dummy angle is no longer close enough to actual angle.

That seems likely.  I don't have any good suggestions, just a few bad ones.

- You could use the same approach as Gromacs, setting a really coarse constraint tolerance so it just won't try very hard to enforce them.  This will lead to bad energy conservation.
- You could replace one or more of the constraints with a harmonic restraint.  That should make the constraint problem easier so it can resolve the remaining ones.  If you set a strong force constant for the restraint, the distance will remain very close to the target but it may require using a smaller time step.  A weaker force constant will allow a larger time step, but the distance will fluctuate more.

**Comment by jlmaccal** [comment]
I was afraid you were going to say that...

Ideally, we would like to be able to reproduce the energies and forces exactly without modification of the topology. And to run stably with similar time steps.

We can achieve that already, except for cholesterol, which is causing many difficulties due to its use of constraints and vsites. This particularly topology is also a problem for LINCS, where it requires a very high order to avoid artificial temperature gradients.

**Comment by peastman** [comment]
Yeah, constraints are really painful.  Conceptually they seem almost the same as restraints, but mathematically they're way more complicated.  They make just about anything you do much harder, whether it's integration, optimization, or even just loading a structure and computing its energy.
